src/qmapshack/poi/CPoiPOI.cpp

```diff
@@ -86,126 +88,269 @@ void CPoiPOI::draw(IDrawContext::buffer_t& buf)
+        {
+            posPoiHighlight.append(poiGroup.iconCenter);
+        }
+    }
 }
 
 bool CPoiPOI::getToolTip(const QPoint &px, QString &str) const
 {
-    poi_t poiFound;
-    bool success = findPoiCloseBy(px, poiFound);
+    poiGroup_t poiGroup;
+    bool success = getPoiGroupCloseBy(px, poiGroup);
     if(success)
     {
-        str += "<b>" + poiFound.name + "</b><br>";
-        str += poiFound.desc;
+        if(poiGroup.pois.count() == 1)
+        {
+            const CRawPoi& poiFound = loadedPois[*poiGroup.pois.begin()];
+            const QString& name = poiFound.getName(false);
+            if(!name.isEmpty())
+            {
+                str += "<b>" + name + "</b><br>\n";
+            }
+            str += tr("From category <b>%1</b><br>\n").arg(poiFound.getCategory());
+            str += poiFound.getData().join("<br>\n");
+        }
+        else
+        {
+            str += tr("<i>Zoom in to see more details.</i>");
+            if(poiGroup.pois.count() <= 10)
+            {
+                str += tr("<br>\nPOIs at this point:");
+                for(quint64 poiID : poiGroup.pois)
+                {
+                    str += "<br>\n<b>" + loadedPois[poiID].getName() + "</b>";
+                }
+            }
+        }
     }
     return success;
 }
 
 void CPoiPOI::addTreeWidgetItems(QTreeWidget* widget)
 {
-    QMap<QString, CPoiCategory*> parentMap;
+    // Open database here so it belongs to the right thread
+    if(!QSqlDatabase::contains(filename + "_widget"))
+    {
+        QSqlDatabase db = QSqlDatabase::addDatabase("QSQLITE", filename + "_widget");
+        db.setDatabaseName(filename);
+        if(!db.open())
+        {
+            qDebug() << "failed to open database" << db.lastError();
+            isActivated = false;
+            return;
+        }
+    }
 
-    for(const QString& category : tagMap.keys())
+    QSqlQuery query("SELECT id, name, parent FROM main.poi_categories ORDER BY id DESC", QSqlDatabase::database(filename + "_widget"));
+
+    QMap<uint, CPoiCategory*> categoryMap;
+
+    while (query.next())
     {
-        QStringList parentChildNameList = category.split("_");
-        QString parentName = parentChildNameList[0];
-        QString childName = parentChildNameList.mid(1).join("_");
-        if(!parentMap.contains(parentName))
+        bool hasParent = false;
+        quint64 parentID = query.value(eSqlColumnCategoryParent).toUInt(&hasParent);
+        if(!hasParent)
+        {
+            continue; //This is the root item
+        }
+        QString categoryName = query.value(eSqlColumnCategoryName).toString();
+        quint64 categoryID = query.value(eSqlColumnCategoryId).toUInt();
+        categoryNames[categoryID] = categoryName;
+        // This assumes, that the parent categories always have a higher ID than their children. This seems to be the case for all Mapsforge poi files
+        if (categoryMap.keys().contains(parentID))
```

Comment:
This might get expensive gathering and searching all the keys over and over again with increasing number. But I have no better idea right now.

Reply:
Using contains directly on the map instead of extracting a list should be O(1)